controllers/utils/cleanup.go

```diff
@@ -57,15 +63,168 @@ func CleanupObjects(ctx context.Context, c client.Client,
 	}
 	l.Info("deleting temporary objects")
 	for _, obj := range types {
-		err := c.DeleteAllOf(ctx, obj, options...)
-		if client.IgnoreNotFound(err) != nil {
-			l.Error(err, "unable to delete object(s)")
+		_, ok := obj.(*snapv1.VolumeSnapshot)
+		if ok {
+			// Handle volumesnapshots differently - do not delete if they have a specific label
+			err := cleanupSnapshots(ctx, c, logger, owner)
+			if err != nil {
+				l.Error(err, "unable to delete volume snapshot(s)")
+				return err
+			}
+		} else {
+			err := c.DeleteAllOf(ctx, obj, options...)
+			if client.IgnoreNotFound(err) != nil {
+				l.Error(err, "unable to delete object(s)")
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Could be generalized to other types if we want to use unstructuredList - would need to pass in group, version, kind
+func cleanupSnapshots(ctx context.Context, c client.Client,
+	logger logr.Logger, owner client.Object) error {
+	// Load current list of snapshots with the cleanup label
+	listOptions := []client.ListOption{
+		client.MatchingLabels{cleanupLabelKey: string(owner.GetUID())},
+		client.InNamespace(owner.GetNamespace()),
+	}
+	snapList := &snapv1.VolumeSnapshotList{}
+	err := c.List(ctx, snapList, listOptions...)
+	if err != nil {
+		return err
+	}
+
+	return CleanupSnapshotsWithLabelCheck(ctx, c, logger, owner, snapList)
+}
+
+func CleanupSnapshotsWithLabelCheck(ctx context.Context, c client.Client,
+	logger logr.Logger, owner client.Object, snapList *snapv1.VolumeSnapshotList) error {
+	// If marked as do-not-delete, remove the cleanup label and ownership
+	snapsForCleanup, err := relinquishSnapshotsWithDoNotDeleteLabel(ctx, c, logger, owner, snapList)
+	if err != nil {
+		return err
+	}
+
+	// Remaining snapshots should be cleaned up
+	for i := range snapsForCleanup {
+		snapForCleanup := &snapList.Items[i]
+
+		// Use a delete precondition to avoid timing issues.
+		// If the object was modified (for example by someone adding a new label) in-between us loading it and
+		// performing the delete, the should throw an error as the resourceVersion will not match
+		snapResourceVersion := snapForCleanup.ResourceVersion
+
+		deleteOptions := &client.DeleteOptions{
+			Preconditions: &metav1.Preconditions{
+				ResourceVersion: &snapResourceVersion,
+			},
+		}
+
+		err := c.Delete(ctx, snapForCleanup, deleteOptions)
```

Comment:
Can be simplified to:
```suggestion
		err := c.Delete(ctx, snapForCleanup, client.Preconditions{ResourceVersion: &snapForCleanup.ResourceVersion})
```